File: frontend/src/types/dualCultivation.ts
/**
 * M7 L7 双修 / 时长榜类型。
 */

export type DualGender = 'male' | 'female'
export type DualBondKind = 'companion' | 'vessel'
export type DualRole = 'number_one' | 'zero'

export interface DualTechnique {
  technique_id: string
  label: string
  mode: 'mutual_gain' | 'transfer' | 'extract'
  mode_label_zh?: string
  require_opposite_gender?: boolean
  base_yield?: number
  [key: string]: unknown
}

export interface DualInviteTarget {
  bond_id: number
  bond_kind: DualBondKind | string
  peer_character_id: number
  peer_name: string
  status: string
  online?: boolean
  peer_major_realm_name?: string | null
  peer_cultivation_points?: number
}

export interface DualSessionParty {
  character_id: number
  name: string
  gender?: DualGender | null
}

export interface DualDiceSnapshot {
  roll?: number | null
  lo?: number | null
  hi?: number | null
  effect_tier?: string | null
  yield_mult?: number | null
  duration_sec?: number | null
  label_zh?: string | null
  rerolls_used?: number
  purpose?: string
  max_rerolls?: number
}

export interface DualSession {
  session_id: number
  status: string
  status_label_zh?: string
  technique_id: string
  technique_label?: string
  mode?: string
  bond_kind?: DualBondKind | string | null
  inviter_role?: DualRole | string
  auto_forced?: boolean
  inviter: DualSessionParty
  invitee: DualSessionParty
  invite_expire_at?: string | null
  undress_expire_at?: string | null
  invitee_undressed?: boolean
  can_undress?: boolean
  can_start?: boolean
  dice?: DualDiceSnapshot | null
  settle_summary?: Record<string, unknown> | null
}

export interface DualMePayload {
  gender: DualGender | null
  gender_label_zh?: string
  needs_gender: boolean
  session: DualSession | null
  techniques: DualTechnique[]
  invite_targets?: {
    companions: DualInviteTarget[]
    vessels: DualInviteTarget[]
    vessel_invite_enabled?: boolean
    hint_zh?: string
  }
  config?: {
    invite_expire_sec?: number
    undress_expire_sec?: number
    max_rerolls?: number
    spirit_stone_cost?: number
    cultivation_gap_scale?: number
  }
}

export interface DualRankEntry {
  rank: number
  character_id: number
  name: string
  score: number
  gender?: string
  score_unit_zh?: string
}

export interface DualBoardPayload {
  board_key: string
  label_zh: string
  min_score: number
  entries: DualRankEntry[]
  my_rank: number | null
  my_score: number
  score_unit_zh?: string
}

export interface DualRanksPayload {
  boards: Record<string, DualBoardPayload>
  my_gender?: DualGender | null
  primary_board?: string
}

File: backend/app/schemas/dual_cultivation.py
"""双修 HTTP DTO（M7 L7）。"""

from __future__ import annotations

from pydantic import BaseModel, Field


class DualInviteRequest(BaseModel):
    """发起双修邀约（仅道侣/炉鼎 + 角色 id）。"""

    technique_id: str = Field(description="双修功法 id")
    target_character_id: int = Field(description="对方角色 id")
    bond_kind: str = Field(description="companion | vessel")
    inviter_role: str | None = Field(
        default="number_one",
        description="邀请方角色位 number_one|zero，默认主动",
    )
    # 兼容旧客户端：若传入则服务端拒绝
    target_name: str | None = None
    # 测用可选种子；线上可忽略
    dice_seed: int | None = None


class DualSetGenderRequest(BaseModel):
    """存量角色一次性补选性别。"""

    gender: str = Field(description="male | female")


class DualRollRequest(BaseModel):
    """掷骰（可选强制种子，仅测/GM）。"""

    dice_seed: int | None = None
